scd/usbtoken.c

```diff
@@ -0,0 +1,737 @@
+	mem_free(s);
+	return -1;
+}
+
+/**
+ * initializes the ctapi interface to the token reader.
+ * If the reader is unplugged during runtime this will ensure it is either
+ * reaquired using the cirrect usb port or not at all
+ */
+static int
+usbtoken_init_ctapi_int(usbtoken_t *token)
+{
+	TRACE("USBTOKEN: usbtoken_init_ctapi_int");
+
+	ASSERT(token);
+
+	int rc;
+	unsigned short lr, port;
+	unsigned char *readers = mem_alloc0(MAX_CT_READERS_SIZE);
+	if (!readers) {
+		ERROR("Failed to allocate memoryfor ct_reader");
+		mem_free(readers);
+		return -1;
+	}
+
+	lr = MAX_CT_READERS_SIZE;
+	CT_list(readers, &lr, 0);
+
+	if (lr <= 0) {
+		ERROR("No usb token reader found.");
+		return -1;
+	}
+
+	rc = token_filter_by_serial(readers, lr, token->serial, &port);
+	if (rc != 0) {
+		ERROR("Could not find specified token reader with serial %s", token->serial);
+		return -1;
+	}
+	token->port = port;
+
+	rc = CT_init(token->ctn, port);
+	if (rc != 0) {
+		ERROR("USBTOKEN: Token reader initialization failed. Ret code: %d", rc);
+		return -1;
+	}
+
+	requestICC(token->ctn);
+
+	rc = queryPIN(token->ctn);
+	if (rc != 0x9000) {
+		selectHSM(token->ctn);
+		rc = queryPIN(token->ctn);
+		TRACE("usbtoken_init queryPIN: %d", rc);
+	}
+
+	return 0;
+}
+
+/**
+ * Initializes a usb token
+ */
+usbtoken_t *
+usbtoken_init(const char *serial)
+{
+	TRACE("USBTOKEN: usbtoken_init");
+
+	ASSERT(serial);
+
+	int rc;
+	usbtoken_t *token = NULL;
+
+	token = mem_new0(usbtoken_t, 1);
+	ASSERT(token);
+
+	token->locked = true;
+	token->ctn = g_ctn++;
+	token->serial = mem_strdup(serial);
+	if (!token->serial) {
+		ERROR("USBTOKEN: Allocating meory for token uuid failed");
+		NULL;
+	}
+
+	rc = usbtoken_init_ctapi_int(token);
+	if (rc != 0) {
+		ERROR("Failed to initialize ctapi interface to usb token reader");
+		mem_free(token);
+		return NULL;
+	}
+
+	token->locked = true;
+	token->serial = mem_strdup(serial);
+
+	CT_close(token->ctn);
+
+	TRACE("Usbtoken initialized");
+	return token;
+}
+
+static int
+provision_auth_code(usbtoken_t *token, const char *tpin, const char *newpass,
+		    unsigned char *pairing_secret, size_t pairing_sec_len)
+{
+	TRACE("USBTOKEN: provision_auth_code");
+
+	ASSERT(token);
+	ASSERT(tpin);
+	ASSERT(newpass);
+	ASSERT(pairing_secret);
+
+	int rc;
+	int newcode_len;
+	unsigned char newcode[TOKEN_MAX_AUTH_CODE_LEN];
+
+	rc = usbtoken_init_ctapi_int(token);
+	if (rc != 0) {
+		ERROR("Failed to initialize ctapi interface to usb token reader");
+		rc = -1;
+		goto out;
+	}
+
+	rc = queryPIN(token->ctn);
+	if (rc != 0x6984) {
+		ERROR("USBTOKEN: Pin is not in tranport state. Aborting");
+		rc = -1;
+		goto out;
+	}
+
+	newcode_len =
+		get_auth_code(newpass, pairing_secret, pairing_sec_len, newcode, sizeof(newcode));
+	if (newcode_len < 0) {
+		ERROR("Could not derive new authentication code");
+		rc = -1;
+		goto out;
+	}
+
+	rc = changePIN(token->ctn, (unsigned char *)tpin, strlen(tpin), newcode, newcode_len);
+
+	memset(newcode, 0, sizeof(newcode));
```

Comment:
adapt if switching to heap allocation

Reply:
i'd not switch as the auth code is only 16B long for now